File: engine/filename.py
from __future__ import annotations

import re
from pathlib import Path

DEFAULT_OUTPUT_FILENAME = "assembled.mp4"
_INVALID_CHARS = re.compile(r'[<>:"/\\|?*\x00-\x1f]')


def sanitize_output_filename(raw: str | None) -> str:
    trimmed = (raw or "").strip()
    if not trimmed:
        return DEFAULT_OUTPUT_FILENAME

    safe = _INVALID_CHARS.sub("", trimmed)
    safe = re.sub(r"\s+", " ", safe).strip()
    if not safe:
        return DEFAULT_OUTPUT_FILENAME

    path = Path(safe)
    stem = path.stem or "assembled"
    return f"{stem}.mp4"

File: lib/assembler/filename.ts
const DEFAULT_OUTPUT_FILENAME = "assembled.mp4";

/** Normalize a user-provided episode filename for MP4 output. */
export function normalizeOutputFilename(raw: string | undefined | null): string {
  const trimmed = (raw ?? "").trim();
  if (!trimmed) {
    return DEFAULT_OUTPUT_FILENAME;
  }

  const safe = trimmed
    .replace(/[<>:"/\\|?*\x00-\x1f]/g, "")
    .replace(/\s+/g, " ")
    .trim();

  if (!safe) {
    return DEFAULT_OUTPUT_FILENAME;
  }

  return safe.toLowerCase().endsWith(".mp4") ? safe : `${safe}.mp4`;
}